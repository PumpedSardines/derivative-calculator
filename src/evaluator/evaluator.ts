import type { Node } from "../types";

export function evaluate(node: Node): Node {
  let result = recursiveEvaluate(node);
  result = reconstructNegative(result);
  result = generateDepends(result);
  return result;
}

function generateDepends(node: Node): Node {
  if (node.type === "variable") {
    return {
      ...node,
      depends: [node.value],
    };
  }

  if ("args" in node) {
    const args = node.args.map(generateDepends) as [Node, Node];
    return {
      ...node,
      args,
      depends: [...new Set(args.flatMap((arg) => arg.depends))],
    };
  }

  if ("arg" in node) {
    const arg = generateDepends(node.arg);
    return {
      ...node,
      arg,
      depends: structuredClone(arg.depends),
    };
  }

  return node;
}

function reconstructNegative(node: Node): Node {
  if (node.type === "number" && node.parsed < 0) {
    return {
      type: "neg",
      depends: [],
      arg: {
        depends: [],
        type: "number",
        parsed: -node.parsed,
        value: `${-node.parsed}`,
      },
    };
  }

  if ("args" in node) {
    return {
      ...node,
      args: node.args.map(reconstructNegative) as [Node, Node],
    };
  }

  if ("arg" in node) {
    return {
      ...node,
      arg: reconstructNegative(node.arg),
    };
  }

  return node;
}

function recursiveEvaluate(node: Node): Node {
  switch (node.type) {
    case "+": {
      const left = recursiveEvaluate(node.args[0]);
      const right = recursiveEvaluate(node.args[1]);

      if (nodeNumberEq(left, 0)) {
        return right;
      }
      if (nodeNumberEq(right, 0)) {
        return left;
      }

      return evalArgsOperator(node, (a, b) => a + b);
    }
    case "-": {
      const left = recursiveEvaluate(node.args[0]);
      const right = recursiveEvaluate(node.args[1]);

      if (right.type === "neg") {
        return evaluate({
          type: "+",
          depends: [],
          args: [left, right.arg],
        });
      }

      if (right.type === "number" && right.parsed < 0) {
        return evaluate({
          type: "+",
          depends: [],
          args: [
            left,
            {
              type: "neg",
              depends: [],
              arg: right,
            },
          ],
        });
      }

      return evalArgsOperator(node, (a, b) => a - b);
    }
    case "*": {
      const left = recursiveEvaluate(node.args[0]);
      const right = recursiveEvaluate(node.args[1]);

      if (nodeNumberEq(left, 0) || nodeNumberEq(right, 0)) {
        return {
          depends: [],
          type: "number",
          parsed: 0,
          value: "0",
        };
      }

      if (nodeNumberEq(left, 1)) {
        return right;
      }
      if (nodeNumberEq(right, 1)) {
        return left;
      }

      if (left.type === "/") {
        return evaluate({
          type: "/",
          depends: [],
          args: [
            {
              type: "*",
              depends: [],
              args: [left.args[0], right],
            },
            left.args[1],
          ],
        });
      }

      if (right.type === "/") {
        return evaluate({
          type: "/",
          depends: [],
          args: [
            {
              type: "*",
              depends: [],
              args: [right.args[0], left],
            },
            right.args[1],
          ],
        });
      }

      return evalArgsOperator(node, (a, b) => a * b);
    }
    case "/":
      return evalArgsOperator(node, (a, b) => {
        if (b === 0) throw new Error("Cannot divide by zero");
        const v = a / b;
        return v === Math.floor(v) ? v : undefined;
      });
    case "^": {
      const b = recursiveEvaluate(node.args[1]);
      if ("parsed" in b && b.parsed === 1) {
        return recursiveEvaluate(node.args[0]);
      }

      return evalArgsOperator(node, (a, b) => {
        const v = Math.pow(a, b);
        if (v === Math.floor(v) && v < 1e3) {
          return v;
        }
        return undefined;
      });
    }
    case "neg": {
      const val = recursiveEvaluate(node.arg);
      if (val.type === "number") {
        return {
          depends: [],
          type: "number",
          parsed: -val.parsed,
          value: `${-val.parsed}`,
        };
      }
      if (val.type === "neg") {
        return val.arg;
      }
      return {
        ...node,
        arg: val,
      };
    }
    case "exp": {
      const arg = recursiveEvaluate(node.arg);
      if (arg.type === "ln") {
        return arg.arg;
      }
      return {
        ...node,
        arg,
      };
    }
    case "ln":
      return {
        ...node,
        arg: recursiveEvaluate(node.arg),
      };
    case "sqrt":
      return evalArgOperator(node, (a) => {
        if (a >= 0) {
          const v = Math.sqrt(a);
          return v === Math.floor(v) ? v : undefined;
        }

        throw new Error("Cannot take square root of negative number");
      });

    case "number":
    case "constant":
    case "variable":
      return node;
  }

  return node;
}

function nodeNumberEq<T extends number>(
  node: Node,
  number: T,
): node is {
  type: "number";
  parsed: T;
  value: `${T}`;
  depends: string[];
} {
  return node.type === "number" && node.parsed === number;
}

function evalArgOperator(
  node: Node,
  cb: (a: number) => number | undefined,
): Node {
  if (!("arg" in node)) {
    throw new Error("Expected node to have args");
  }

  const value = recursiveEvaluate(node.arg);

  if (value.type === "number") {
    const parsed = cb(value.parsed);
    // Means that the numbers are not valid
    if (parsed == null) return node;
    return {
      depends: [],
      type: "number",
      parsed,
      value: parsed.toString(),
    };
  }

  return node;
}

function evalArgsOperator(
  node: Node,
  cb: (a: number, b: number) => number | undefined,
): Node {
  if (!("args" in node)) {
    throw new Error("Expected node to have args");
  }

  const left = recursiveEvaluate(node.args[0]);
  const right = recursiveEvaluate(node.args[1]);

  if (left.type === "number" && right.type === "number") {
    const parsed = cb(left.parsed, right.parsed);
    // Means that the numbers are not valid
    if (parsed == null) return node;
    return {
      depends: [],
      type: "number",
      parsed,
      value: parsed.toString(),
    };
  }

  return {
    ...node,
    args: [left, right],
  };
}
