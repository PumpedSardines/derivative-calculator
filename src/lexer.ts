export const elementaryOperators = ["^", "*", "/", "+", "-"] as const;
export const functionOperators = [
  "sin",
  "cos",
  "tan",
  "ln",
  "arcsin",
  "arccos",
  "arctan",
  "sqrt",
] as const;
const constants = ["pi", "e"] as const;

export type Operator =
  | (typeof elementaryOperators)[number]
  | (typeof functionOperators)[number];
export type Constant = (typeof constants)[number];

type ParsingType = "none" | "operator" | "number" | "variable";

export type Token =
  | { type: Operator }
  | { type: "(" | ")" }
  | {
    type: "variable";
    value: string;
  }
  | {
    type: "constant";
    value: Constant;
  }
  | { type: "number"; value: string; parsed: number };

/**
 * Tokenizes the input string into an array of mathematical tokens.
 *
 * @param input The input string to tokenize.
 * @returns An array of tokens.
 * @throws If the input contains invalid characters.
 */
export function tokenize(input: string): Token[] {
  const tokens: Token[] = [];

  let parsingType: ParsingType = "none";
  let currentValue = "";
  let wasLastEInNumber = false;

  const chars = input.split("");
  let index = 0;

  const pushVariable = (value: string) => {
    if (functionOperators.includes(value as any)) {
      tokens.push({ type: value as Operator });
    } else if (constants.includes(value as Constant)) {
      tokens.push({ type: "constant", value: value as Constant });
    } else {
      tokens.push({ type: "variable", value });
    }
  };

  const pushNumber = (value: string) => {
    if (/^\d+(\.\d+)?(e-?\d+)?$/.test(value) === false) {
      throw new Error(`Invalid number: ${value}`);
    }

    tokens.push({ type: "number", value, parsed: Number(value) });
  };

  letterLoop: while (index < chars.length) {
    const char = chars[index]!;

    if ([" ", "\n", "\t"].includes(char)) {
      index++;

      if (parsingType == "variable") {
        pushVariable(currentValue);
        parsingType = "none";
        currentValue = "";
      }

      if (parsingType == "number") {
        pushNumber(currentValue);
        parsingType = "none";
        currentValue = "";
      }

      continue;
    }

    if (parsingType == "none") {
      parsingType = getParsingTypeFromValue(char);

      switch (parsingType) {
        case "variable":
        case "number":
          currentValue = char;
          index++;
          break;
        case "operator":
          tokens.push({ type: char as Operator });
          parsingType = "none";
          index++;
          break;
      }

      continue letterLoop;
    }

    if (parsingType == "variable") {
      if (getParsingTypeFromValue(char) == "variable") {
        currentValue += char;
      } else {
        pushVariable(currentValue);
        parsingType = "none";
        currentValue = "";
        continue letterLoop;
      }
    }

    if (parsingType == "number") {
      if (char === "-" && wasLastEInNumber) {
        wasLastEInNumber = false;
        currentValue += char;
      } else if (getParsingTypeFromValue(char) == "number" || char === "e") {
        if (char === "e") {
          wasLastEInNumber = true;
        } else {
          wasLastEInNumber = false;
        }
        currentValue += char;
      } else {
        pushNumber(currentValue);
        parsingType = "none";
        currentValue = "";
        wasLastEInNumber = false;
        continue letterLoop;
      }
    }

    index++;
  }

  if (parsingType == "variable") {
    pushVariable(currentValue);
  }

  if (parsingType == "number") {
    pushNumber(currentValue);
  }

  return tokens;
}

function getParsingTypeFromValue(value: string): ParsingType {
  if (!isNaN(Number(value)) || value == ".") {
    return "number";
  } else if (
    elementaryOperators.includes(value as any) ||
    value == "(" ||
    value == ")"
  ) {
    return "operator";
  } else if (value.match(/[a-zA-Z]/)) {
    return "variable";
  }

  throw new Error(`Invalid character: ${value}`);
}
