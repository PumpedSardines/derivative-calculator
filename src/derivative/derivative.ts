import { evaluate } from "../evaluator/evaluator";
import { Node } from "../types";

type DerivationOptions = {
  variable: string;
};

export function derivative(node: Node): Node {
  let result = recursiveDerivative(node, { variable: "x" });
  result = evaluate(result);
  return result;
}

function recursiveDerivative(node: Node, options: DerivationOptions): Node {
  const variable = options.variable;

  if (!node.depends.includes(variable)) {
    return { depends: [], type: "number", value: "0", parsed: 0 };
  }

  if (node.type === "variable") {
    return { depends: [], type: "number", value: "1", parsed: 1 };
  }

  if (node.type === "neg") {
    return {
      ...node,
      arg: recursiveDerivative(node.arg, options),
    };
  }

  if (node.type === "+") {
    return {
      ...node,
      args: node.args.map((arg) => recursiveDerivative(arg, options)) as [
        Node,
        Node,
      ],
    };
  }

  if (node.type === "-") {
    return {
      ...node,
      args: node.args.map((arg) => recursiveDerivative(arg, options)) as [
        Node,
        Node,
      ],
    };
  }

  if (node.type === "*") {
    const [a, b] = node.args;

    return evaluate({
      type: "+",
      depends: [],
      args: [
        {
          type: "*",
          depends: [],
          args: [recursiveDerivative(a, options), b],
        },
        {
          type: "*",
          depends: [],
          args: [a, recursiveDerivative(b, options)],
        },
      ],
    });
  }

  if (node.type === "^") {
    if (
      node.args[0].depends.includes(variable) &&
      !node.args[1].depends.includes(variable)
    ) {
      return evalPowerRule(node, options);
    }
  }

  if (node.type === "exp") {
    return {
      type: "*",
      depends: [],
      args: [
        {
          type: "exp",
          depends: [],
          arg: node.arg,
        },
        recursiveDerivative(node.arg, options),
      ],
    };
  }

  if (node.type === "ln") {
    return {
      type: "*",
      depends: [],
      args: [
        {
          type: "/",
          depends: [],
          args: [
            {
              type: "number",
              depends: [],
              value: "1",
              parsed: 1,
            },
            node.arg,
          ],
        },
        recursiveDerivative(node.arg, options),
      ],
    };
  }

  if (node.type === "sin") {
    return {
      type: "*",
      depends: [],
      args: [
        {
          type: "cos",
          depends: [],
          arg: node.arg,
        },
        recursiveDerivative(node.arg, options),
      ],
    };
  }

  if (node.type === "cos") {
    return {
      type: "*",
      depends: [],
      args: [
        {
          type: "neg",
          depends: [],
          arg: {
            type: "sin",
            depends: [],
            arg: node.arg,
          },
        },
        recursiveDerivative(node.arg, options),
      ],
    };
  }

  throw new Error("Not implemented");
}

function evalPowerRule(node: Node, options: DerivationOptions): Node {
  if (node.type !== "^") {
    throw new Error("Invalid node type");
  }

  const [a, b] = node.args;

  const final = evaluate({
    type: "*",
    depends: [],
    args: [
      {
        type: "*",
        depends: [],
        args: [
          b,
          {
            type: "^",
            depends: [],
            args: [
              a,
              {
                depends: [],
                type: "-",
                args: [
                  b,
                  {
                    depends: [],
                    type: "number",
                    value: "1",
                    parsed: 1,
                  },
                ],
              },
            ],
          },
        ],
      },
      recursiveDerivative(a, options),
    ],
  });

  return final;
}
