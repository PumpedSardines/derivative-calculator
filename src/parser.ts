import { functionOperators, type Constant, type Token } from "./lexer";

export type Node =
  | {
      type: "+" | "-" | "*" | "/" | "^";
      args: [Node, Node];
      depends: string[];
    }
  | {
      type:
        | "cos"
        | "sin"
        | "tan"
        | "ln"
        | "sqrt"
        | "exp"
        | "arcsin"
        | "arccos"
        | "arctan"
        | "neg";
      arg: Node;
      depends: string[];
    }
  | {
      type: "variable";
      value: string;
      depends: string[];
    }
  | {
      type: "constant";
      value: Constant;
      depends: string[];
    }
  | {
      type: "number";
      value: string;
      parsed: number;
      depends: string[];
    };

type NodeToken = { node: Node } | { token: Token };

export function parse(tokens: Token[]): Node {
  let tokenNodes = evalParentheses(tokens);
  tokenNodes = parseFunctionOperations(tokenNodes);
  let nodes = parseTokenNodeArray(structuredClone(tokenNodes));
  nodes = recursivePass(nodes);
  return nodes;
}

function evalParentheses(tokens: Token[]): NodeToken[] {
  const tokenNodes: NodeToken[] = structuredClone(tokens).map((t) => ({
    token: t,
  }));

  const parentheses: [number, number][] = [];

  for (let i = 0; i < tokens.length; i++) {
    const token = tokens[i]!;

    if (token.type === "(") {
      let depth = 1;
      let j = i + 1;

      while (depth > 0) {
        if (j >= tokens.length) {
          throw new Error("Unmatched parentheses");
        }

        if (tokens[j]!.type === "(") {
          depth++;
        } else if (tokens[j]!.type === ")") {
          depth--;
        }

        j++;
      }

      parentheses.push([i, j - 1]);
      i = j;
    }
  }

  let lengthMin = 0;
  for (const [start, end] of parentheses) {
    const innerTokens = structuredClone(tokens).slice(start + 1, end);
    const innerNode = parse(innerTokens);
    tokenNodes.splice(start - lengthMin, end - start + 1, {
      node: innerNode,
    });
    lengthMin = lengthMin + end - start;
  }

  return tokenNodes;
}

function recursivePass(node: Node): Node {
  if (
    node.type === "^" &&
    node.args[0].type === "constant" &&
    node.args[0].value === "e"
  ) {
    return recursivePass({
      type: "exp",
      arg: node.args[1],
      depends: [],
    });
  }

  if ("args" in node) {
    const args = node.args.map(recursivePass) as [Node, Node];
    return {
      type: node.type,
      args,
      depends: [],
    };
  } else if ("arg" in node) {
    const arg = recursivePass(node.arg);
    return {
      type: node.type,
      arg,
      depends: [],
    };
  }

  return node;
}

function parseFunctionOperations(nodeTokens: NodeToken[]): NodeToken[] {
  nodeTokens = structuredClone(nodeTokens);

  for (let i = nodeTokens.length - 2; i >= 0; i--) {
    const nodeToken = nodeTokens[i]!;
    if ("node" in nodeToken) {
      continue;
    }
    const token = nodeToken.token;

    if (functionOperators.includes(token.type as any)) {
      const node = parseTokenNodeArray([nodeTokens[i + 1]!]);
      nodeTokens.splice(i, 2, {
        node: {
          type: token.type as (typeof functionOperators)[number],
          arg: node,
          depends: [],
        },
      });

      return parseFunctionOperations(nodeTokens);
    }
  }

  return nodeTokens;
}

function parseTokenNodeArray(nodeTokens: NodeToken[]): Node {
  nodeTokens = structuredClone(nodeTokens);

  if (nodeTokens.length === 0) {
    throw new Error("Invalid expression");
  }

  if (nodeTokens.length === 1) {
    const nodeToken = nodeTokens[0]!;

    if ("node" in nodeToken) {
      return nodeToken.node;
    } else {
      const token = nodeToken.token;

      if (token.type === "number") {
        return {
          type: "number",
          value: token.value,
          parsed: token.parsed,
          depends: [],
        };
      } else if (token.type === "variable") {
        return {
          type: "variable",
          value: token.value,
          depends: [],
        };
      } else if (token.type === "constant") {
        return {
          type: "constant",
          value: token.value,
          depends: [],
        };
      }
    }
  }

  for (const operatorPairs of [
    ["+", "-"],
    ["*", "/"],
    ["neg"],
    ["^"],
  ] as const) {
    if (operatorPairs[0] === "neg") {
      const value = (() => {
        const nodeToken = nodeTokens[0]!;

        if ("node" in nodeToken) {
          return;
        }

        const token = nodeToken.token;

        if (token.type === "-") {
          const node = parseTokenNodeArray(nodeTokens.slice(1));
          return {
            type: "neg",
            arg: node,
            depends: [] as string[],
          } as const;
        }

        return;
      })();

      if (value) return value;
    }

    for (let i = nodeTokens.length - 1; i >= 1; i--) {
      const nodeToken = nodeTokens[i]!;

      if ("node" in nodeToken) {
        continue;
      }

      const token = nodeToken.token;

      if (token.type === "-") {
        const prevNodeToken = nodeTokens[i - 1]!;

        if ("token" in prevNodeToken) {
          const prevToken = prevNodeToken.token;

          if (!["constant", "variable", "number"].includes(prevToken.type)) {
            continue;
          }
        }
      }

      if (token.type === operatorPairs[0] || token.type === operatorPairs[1]) {
        return {
          type: token.type,
          depends: [],
          args: [
            parseTokenNodeArray(nodeTokens.slice(0, i)),
            parseTokenNodeArray(nodeTokens.slice(i + 1)),
          ],
        };
      }
    }
  }

  throw new Error("Invalid expression");
}
