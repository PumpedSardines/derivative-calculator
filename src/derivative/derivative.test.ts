import { expect, it } from "@jest/globals";
import { tokenize } from "../lexer/lexer";
import { parse } from "../parser/parser";
import { evaluate } from "../evaluator/evaluator";
import { derivative } from "./derivative";

const compile = (input: string) => evaluate(parse(tokenize(input)));

it.each(["x^2", "(x + 8)^5", "x^(sin(2) - pi)"])(
  "derives power rule %s",
  (input) => {
    const result = derivative(compile(input));
    expect(result).toMatchSnapshot();
  },
);

it.each(["e^(2 * x)"])("derives exp %s", (input) => {
  const result = derivative(compile(input));
  expect(result).toMatchSnapshot();
});

it.each(["ln(x)"])("derives logarithms %s", (input) => {
  const result = derivative(compile(input));
  expect(result).toMatchSnapshot();
});

it.each(["sin(x)", "-cos(x)", "ln(sin(x))"])("derives trig %s", (input) => {
  const result = derivative(compile(input));
  expect(result).toMatchSnapshot();
});
