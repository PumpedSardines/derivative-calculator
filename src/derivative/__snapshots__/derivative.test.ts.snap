// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`derives exp e^(2 * x) 1`] = `
{
  "args": [
    {
      "arg": {
        "args": [
          {
            "depends": [],
            "parsed": 2,
            "type": "number",
            "value": "2",
          },
          {
            "depends": [
              "x",
            ],
            "type": "variable",
            "value": "x",
          },
        ],
        "depends": [
          "x",
        ],
        "type": "*",
      },
      "depends": [
        "x",
      ],
      "type": "exp",
    },
    {
      "depends": [],
      "parsed": 2,
      "type": "number",
      "value": "2",
    },
  ],
  "depends": [
    "x",
  ],
  "type": "*",
}
`;

exports[`derives logarithms ln(x) 1`] = `
{
  "args": [
    {
      "depends": [],
      "parsed": 1,
      "type": "number",
      "value": "1",
    },
    {
      "depends": [
        "x",
      ],
      "type": "variable",
      "value": "x",
    },
  ],
  "depends": [
    "x",
  ],
  "type": "/",
}
`;

exports[`derives power rule (x + 8)^5 1`] = `
{
  "args": [
    {
      "depends": [],
      "parsed": 5,
      "type": "number",
      "value": "5",
    },
    {
      "args": [
        {
          "args": [
            {
              "depends": [
                "x",
              ],
              "type": "variable",
              "value": "x",
            },
            {
              "depends": [],
              "parsed": 8,
              "type": "number",
              "value": "8",
            },
          ],
          "depends": [
            "x",
          ],
          "type": "+",
        },
        {
          "depends": [],
          "parsed": 4,
          "type": "number",
          "value": "4",
        },
      ],
      "depends": [
        "x",
      ],
      "type": "^",
    },
  ],
  "depends": [
    "x",
  ],
  "type": "*",
}
`;

exports[`derives power rule x^(sin(2) - pi) 1`] = `
{
  "args": [
    {
      "args": [
        {
          "arg": {
            "depends": [],
            "parsed": 2,
            "type": "number",
            "value": "2",
          },
          "depends": [],
          "type": "sin",
        },
        {
          "depends": [],
          "type": "constant",
          "value": "pi",
        },
      ],
      "depends": [],
      "type": "-",
    },
    {
      "args": [
        {
          "depends": [
            "x",
          ],
          "type": "variable",
          "value": "x",
        },
        {
          "args": [
            {
              "args": [
                {
                  "arg": {
                    "depends": [],
                    "parsed": 2,
                    "type": "number",
                    "value": "2",
                  },
                  "depends": [],
                  "type": "sin",
                },
                {
                  "depends": [],
                  "type": "constant",
                  "value": "pi",
                },
              ],
              "depends": [],
              "type": "-",
            },
            {
              "depends": [],
              "parsed": 1,
              "type": "number",
              "value": "1",
            },
          ],
          "depends": [],
          "type": "-",
        },
      ],
      "depends": [
        "x",
      ],
      "type": "^",
    },
  ],
  "depends": [
    "x",
  ],
  "type": "*",
}
`;

exports[`derives power rule x^2 1`] = `
{
  "args": [
    {
      "depends": [],
      "parsed": 2,
      "type": "number",
      "value": "2",
    },
    {
      "depends": [
        "x",
      ],
      "type": "variable",
      "value": "x",
    },
  ],
  "depends": [
    "x",
  ],
  "type": "*",
}
`;

exports[`derives trig -cos(x) 1`] = `
{
  "arg": {
    "depends": [
      "x",
    ],
    "type": "variable",
    "value": "x",
  },
  "depends": [
    "x",
  ],
  "type": "sin",
}
`;

exports[`derives trig ln(sin(x)) 1`] = `
{
  "args": [
    {
      "arg": {
        "depends": [
          "x",
        ],
        "type": "variable",
        "value": "x",
      },
      "depends": [
        "x",
      ],
      "type": "cos",
    },
    {
      "arg": {
        "depends": [
          "x",
        ],
        "type": "variable",
        "value": "x",
      },
      "depends": [
        "x",
      ],
      "type": "sin",
    },
  ],
  "depends": [
    "x",
  ],
  "type": "/",
}
`;

exports[`derives trig sin(x) 1`] = `
{
  "arg": {
    "depends": [
      "x",
    ],
    "type": "variable",
    "value": "x",
  },
  "depends": [
    "x",
  ],
  "type": "cos",
}
`;
